### ISSUE TITLE
Support for leaving constructor errors unchanged

### ISSUE THREAD

**Issue by abhinav** [uber-go/dig#145]
dig should support an opt-in way of requesting that errors returned by user
constructors are propagated upwards as-is.

Currently, errors are wrapped with a lot of context:

```
failed to get arguments for [..]:
  couldn't get arguments for constructor [..]:
    constructor [..] failed:
      <actual error>
```

This is all useful in most scenarios except where the error is informative for
the rest of my program. For example, I could have a `parseCLIOptions() (*options, error)`
constructor which parses the command line arguments provided by the user. The
error returned by the function would usually contain the message that should
be printed to the user, but because of the wrapping, the error message is no
longer consumable by users of the CLI application.

I imagine this would take form as a simple Container option:

    dig.New(dig.WrapErrors(false))
    // Name TBD. Another option is WrapUserErrors which is very explicit about
    // what this is doing.
    
Note that this wouldn't apply to all error wrapping inside dig. Only errors
returned by user-provided functions on the Invoke path will be affected.

Note that Invoke already returns the error returned by the provided function
as-is.

    err := c.Invoke(func() error { return someError })
    fmt.Println(err == someError)
    // Output: true

I just want support for,

    c.Provide(func() (*type1, error) { return someError })
    err := c.Invoke(func(*type1) { panic("never called")})
    fmt.Println(err == someError)
    // Output: true

**Comment by abhinav** [comment]
@prashantv suggests considering an alternative approach:

Instead of providing an option, provide a top-level function which returns the
"root cause".

    err := c.Invoke(..)
    fmt.Println(someError == dig.RootCause(err))
    // Output: true

@breerly @akshayjshah what do you think of the two approaches? I've already got the above implemented in #146, and it's trivial to change it to the other.

**Comment by breerly** [comment]
I prefer @prashantv's approach over passing an option that effects the behavior;
with `dig.RootCause`, the cause is always available, it's just a matter of extracting it.
